### ISSUE TITLE
Unsatisfied dependency for Brewtarget update in ubuntu 24.01

### ISSUE THREAD

**Issue by grandpubaa** [Brewtarget/brewtarget#840]
Brewtarget -4.05_24-1 pending on libqt5core5a, which is not available in Ubuntu 24.01. Installation is not possible. Pendency is not available in synaptic.

**Comment by matty0ung** [comment]
Sorry you're hitting problems.  I'll try to look at the dependencies in the packages at some point.  In the meantime, you should be able to workaround by doing the following from the command line before installing Brewtarget:
```
sudo apt install libqt5core5a
```
But give me a shout if this doesn't work.

**Comment by grandpubaa** [comment]
Hi Matt,

Sadly this workaround will not work, it returns the comment: libqt5core5t64 is selected instead off  libqt5core5a

**Comment by matty0ung** [comment]
Ah, hmm, OK.  Maybe we just need to change the name of the dependency.  Could you post the exact error message you get when you try to install Brewtarget?

**Comment by grandpubaa** [comment]
Hi Matt,
Sorry wasn't able to react sooner.
Error:
dependency cannot be fulfilled:
libqt5core5a (>=5.9.5)

![afbeelding](https://github.com/user-attachments/assets/ef1ec143-32ca-43e9-bd74-5e8613dc4675)

**Comment by grandpubaa** [comment]
And if you want to install it through root, or synaptic:
libqt5core5a:

Package libqt5core5a is not available, but exists in the database.
This usually means that the package was either listed as a dependency but never made available, or was removed due to obsolescence, or is not available due to the current source list

![afbeelding](https://github.com/user-attachments/assets/e404c818-2f1c-42d4-9bf6-82d15041c23b)

**Comment by matty0ung** [comment]
Thanks - that's helpful.  Hopefully it's just a question of updating the dependency list in the package.  I'll have a look.

(In the near future, we'll be on Qt 6, but I'm still debugging the upgrade, so I will try to fix this Qt5 dependency in the meantime.)

**Comment by matty0ung** [comment]
When you get a moment, try with the patched installers downloadable from https://github.com/Brewtarget/brewtarget/actions/runs/11132482035 (or tonight's nightly builds when they run) and see if that fixes the problem.  I've replaced the `libqt5core5a` dependency with what I _think_ is its successor library.

**Comment by matty0ung** [comment]
FYI, I think the change in library name was because of efforts in Debian to fix the "year 2038 problem" -- see https://wiki.debian.org/ReleaseGoals/64bit-time.

**Comment by grandpubaa** [comment]
Hi Matt,

The install went fine, although some decencies needed to be removed during installation. Problem is after is says database with new ingredients is successful loaded, it wont start but just close the application. So now I don't have any working version anymore. (24.01 version)

**Comment by matty0ung** [comment]
Ah, no, how frustrating.  Don't worry though.  There will be a solution!  Just have to work out what's causing it to shut down - most likely either an assert or a crash.  Can you post the Brewtarget log file here?

**Comment by grandpubaa** [comment]
[brewtarget_2024_10_02_20_15_20_780.log](https://github.com/user-attachments/files/17247849/brewtarget_2024_10_02_20_15_20_780.log)
[brewtarget_2024_10_02_20_15_34_217.log](https://github.com/user-attachments/files/17247853/brewtarget_2024_10_02_20_15_34_217.log)
[brewtarget_2024_10_02_20_16_02_801.log](https://github.com/user-attachments/files/17247854/brewtarget_2024_10_02_20_16_02_801.log)
[brewtarget_2024_10_02_20_16_31_908.log](https://github.com/user-attachments/files/17247857/brewtarget_2024_10_02_20_16_31_908.log)
[brewtarget_2024_10_02_20_33_47_024.log](https://github.com/user-attachments/files/17247858/brewtarget_2024_10_02_20_33_47_024.log)

**Comment by grandpubaa** [comment]
Hi Matt,

hereby the log files, also have the crash report, but it's 21mb and the file is not supported. So if you want have to convert it to a plain text.

**Comment by matty0ung** [comment]
Thanks for that.  I don't need the crash report for this one, but would you be able to upload your `database.sqlite` file (or email it to me to _matt AT brewtarget DOT beer_).  Hopefully I can then reproduce the problems locally.

**Comment by matty0ung** [comment]
From what I could tell in the log files, the program is encountering unexpected things in the database.  It could be that your database has some unusual data in (that the code didn't account for), or it could be that it somehow got into a bad state (perhaps during one of the DB upgrades).  I'm hoping the remaining work is either to fix your DB correctly or to make improvements to the DB upgrade process so that you can re-run it on your backup copy (which the upgrade process automatically makes before doing any modifications).

**Comment by grandpubaa** [comment]
Hi Matt,

Indeed the problem lays in the data base. As I renamed the data base as old. It start. But every time I try to pull back the old database the program refuse to start.

**Comment by matty0ung** [comment]
Many thanks for sending through your DB file.  It did indeed give a clue to the remaining problem.  

In the old database schema, it's possible to create a salt addition for a recipe that has "never" as the value for "when to add".  This doesn't make a huge amount of sense (unless there is something else you wanted to do with the salt whilst brewing :smiley:), so we omitted this possibility from the new schema.  However, we didn't correctly handle such "never" values when upgrading the database.  I've added a fix (and tested it on your database file) in https://github.com/Brewtarget/brewtarget/pull/856, which should make its way through to the main code within the next week.
